### PR TITLE
Introduce async SourceFileProvider interface

### PR DESCRIPTION
**Pull request by Adjective-Object**
Adds a common interface that in future changes will be implemented
with a faster alternate that avoids scanning all files in the program
during initialization

As part of this work, I rework consumers of TypeScriptProgram to be
asyncronous, and add runWithConcurrentLimit to fence check in parallel.
This is needed because if we were to check all fences in parallel we
would hit the system RLIMIT by opening all fences simultaneously.
I also let runWithConcurrentLimit display a progress bar to give user
feedback while performing fence checks. This does not display during
provider initialization, but this is less of a problem against the (yet
to come) newer implementation that doesn't perform ts program
initialization.

### CODE REVIEW

**Comment by Adjective-Object** [comment]
This change ([commit](https://github.com/Adjective-Object/good-fences/commit/722050c9b7483fc7854e6af946ba8f37ca859c2f ))([branch](https://github.com/Adjective-Object/good-fences/tree/u/mahuangh%2Ffaster-source-providers-3)) that introduces the faster provider is blocked on this change.

File: src/utils/runWithConcurrentLimit.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            },
+            CliProgress.Presets.shades_grey
+        );
+        progressBar.start(inputs.length, 0);
+    }
+
+    const queueNext = (): Promise<void> | void => {
+        const next = i.shift();
+        completedJobs += 1;
+        if (progressBar) {
```
Comment (smikula) [inline src/utils/runWithConcurrentLimit.ts:27]: You should be able to simplify this with optional chaining `progressBar?.(...)`

(And same below.)

**Review by smikula — CHANGES_REQUESTED** [review]
Just one small comment.

File: src/utils/runWithConcurrentLimit.ts (hunk 1 above)
Comment (Adjective-Object) [reply src/utils/runWithConcurrentLimit.ts:27]: My personal preference is to use explicit control flow structures for effects but happy to change this to keep with the project's preferences.

**Pull request merged**